wsmux/session.go

```diff
@@ -260,21 +231,96 @@ func (s *Session) Close() error {
 	return err
 }
 
-// Addr used for implementing net.Listener
+// Addr returns the address of this listener.  This is required for
+// implementing net.Listener, but its return value here is not very useful.
 func (s *Session) Addr() net.Addr {
 	return s.conn.LocalAddr()
 }
 
+// RemoveStream removes the stream with given id
+func (s *Session) RemoveStream(id uint32) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if _, ok := s.streams[id]; ok {
+		delete(s.streams, id)
+	}
+}
+
+// IsClosed returns true if the session is closed.
+func (s *Session) IsClosed() bool {
+	select {
+	case <-s.closed:
+		return true
+	default:
+	}
+	return false
+}
+
+// SetCloseCallback sets the function which is called when the session is closed
+func (s *Session) SetCloseCallback(h func()) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.closeCallback = h
```

Comment:
Indeed, there is quite a bit of extra locking in this app.  I don't think any of it's incorrect, though, and I'm hesitant to remove it lest I introduce some issue.

To this specific point, yeah, I'll remove the defer, but in another PR (since this one isn't changing code).

Reply:
I added this to https://bugzilla.mozilla.org/show_bug.cgi?id=1515530